### PR TITLE
[WIP] Add Pegasus

### LINKED ISSUE #4918: Pegasus for summarization !
# 🌟 New model addition

## Model description

https://ai.googleblog.com/2020/06/pegasus-state-of-art-model-for.html?m=1

https://arxiv.org/abs/1912.08777

Abstract 
Recent work pre-training Transformers with self-supervised objectives on large text corpora has shown great success when fine-tuned on downstream NLP tasks including text summarization. However, pre-training objectives tailored for abstractive text summarization have not been explored. Furthermore there is a lack of systematic evaluation across diverse domains. In this work, we propose pre-training large Transformer-based encoder-decoder models on massive text corpora with a new self-supervised objective. In PEGASUS, important sentences are removed/masked from an input document and are generated together as one output sequence from the remaining sentences, similar to an extractive summary. We evaluated our best PEGASUS model on 12 downstream summarization tasks spanning news, science, stories, instructions, emails, patents, and legislative bills. Experiments demonstrate it achieves state-of-the-art performance on all 12 downstream datasets measured by ROUGE scores. Our model also shows surprising performance on low-resource summarization, surpassing previous state-of-the-art results on 6 datasets with only 1000 examples. Finally we validated our results using human evaluation and show that our model summaries achieve human performance on multiple datasets. 

## Open source status

* [x] the model implementation is available: https://github.com/google-research/pegasus
* [x] the model weights are available: https://github.com/google-research/pegasus
* [x] who are the authors: Jingqing Zhang @JingqingZ, Yao Zhao @yaozhaogoogle, Mohammad Saleh and Peter J. Liu

### PR DESCRIPTION
**Pull request by JingqingZ**
* Add PEGASUS in TF2

### CODE REVIEW

**Comment by JingqingZ** [comment]
Hi @sshleifer, I will continue to push code to this PR and make it runnable asap.

**Comment by JingqingZ** [comment]
@sshleifer The code has been uploaded. (1) The test is runnable in TF2 and loads AESLC checkpoints successfully with correct outputs. (2) Code of models and layers (including decoding, beam search) are all in a single file which may look messy (sorry). (3) Most code is simply copied and pasted (then converted to TF2) from the original PEGASUS repo so you may refer to the original repo for clearer code if necessary. 

I think you can start from here. Please let me know if I can help further.

**Comment by sshleifer** [comment]
@JingqingZ I'm gunna add torch first in #6340 (you will be a PR co-author). And then come back here to finish TF. No action needed from you, just an update.
